haystack/preview/components/samplers/top_p.py

```diff
@@ -114,41 +117,66 @@ def run(self, query: str, documents: List[Document], top_p: Optional[float] = No
+        if needs_scoring:
+            # prepare the data for the model
+            query_doc_pairs = [[query, doc.text] for doc in documents]
+            features = self.tokenizer(query_doc_pairs, padding=True, truncation=True, return_tensors="pt")
 
-        # compute the similarity scores for these combinations
-        similarity_scores = self.cross_encoder.predict(query_doc_pairs)
+            with torch.no_grad():
+                similarity_scores = self.model(**features).logits.squeeze()
+        else:
+            similarity_scores = torch.tensor(self.collect_scores(documents), dtype=torch.float32)
 
         # Apply softmax normalization to the similarity scores
-        probs = np.exp(similarity_scores) / np.sum(np.exp(similarity_scores))
+        probs = torch.exp(similarity_scores) / torch.sum(torch.exp(similarity_scores))
 
         # Sort the probabilities and calculate their cumulative sum
-        sorted_probs = np.sort(probs)[::-1]
-        cumulative_probs = np.cumsum(sorted_probs)
+        sorted_probs, sorted_indices = torch.sort(probs, descending=True)
+        cumulative_probs = torch.cumsum(sorted_probs, dim=0)
 
         # Find the indices with cumulative probabilities that exceed top_p
-        top_p_indices = np.where(cumulative_probs <= top_p)[0]
+        top_p_indices = torch.where(cumulative_probs <= top_p)[0]
 
         # Map the selected indices back to their original indices
-        original_indices = np.argsort(probs)[::-1][top_p_indices]
-        # and select the top_p responses
-        selected_docs = [documents[i] for i in original_indices]
+        original_indices = sorted_indices[top_p_indices]
+        selected_docs = [documents[i.item()] for i in original_indices]
 
-        # low p resulted in no documents being selected, then
+        # If low p resulted in no documents being selected, then
         # return at least one document
         if not selected_docs:
-            highest_prob_indices = np.argsort(probs)[::-1]
-            selected_docs = [documents[highest_prob_indices[0]]]
+            highest_prob_indices = torch.argsort(probs, descending=True)
+            selected_docs = [documents[highest_prob_indices[0].item()]]
 
-        # include prob scores in the results
-        if self.score_field:
+        # Include prob scores in the results
+        if self.score_field and needs_scoring:
             for idx, doc in enumerate(selected_docs):
-                doc.metadata[self.score_field] = str(sorted_probs[idx])
+                doc.metadata[self.score_field] = str(sorted_probs[idx].item())
 
         return {"documents": selected_docs}
+
+    def collect_scores(self, documents: List[Document]) -> List[float]:
+        """
+        Collect the scores from the documents' metadata.
+        :param documents: List of Documents.
+        :return: List of scores.
+        """
+        return [d.metadata[self.score_field] for d in documents]
```

Comment:
This might need an error catch for if the score field is not present in the document. I think it's possible to arrive here when no model_name_or_path is specified and the doc doesn't have scores.

Reply:
```
        if not self.model_name_or_path and not self.has_scores(documents):
            raise ComponentError(
                f"The component {self.__class__.__name__} requires similarity scores in the documents' metadata. "
                "Either set 'model_name_or_path' to score documents or add scores to the documents."
            )
```
Doesn't this block prevent it?

Reply:
Ahh yes you're right, my apologies